staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go

```diff
@@ -76,38 +77,85 @@ const (
+// replaced key Values with the existing ones
 func replaceTopology(pv *v1.PersistentVolume, oldKey, newKey string) error {
+	// Make sure the necessary fields exist
+	if pv == nil || pv.Spec.NodeAffinity == nil || pv.Spec.NodeAffinity.Required == nil ||
+		pv.Spec.NodeAffinity.Required.NodeSelectorTerms == nil || len(pv.Spec.NodeAffinity.Required.NodeSelectorTerms) == 0 {
+		return nil
+	}
 	for i := range pv.Spec.NodeAffinity.Required.NodeSelectorTerms {
+		newKeyValsSet := make(map[string]bool)
+		var requirementsToRemove []int
 		for j, r := range pv.Spec.NodeAffinity.Required.NodeSelectorTerms[i].MatchExpressions {
 			if r.Key == oldKey {
 				pv.Spec.NodeAffinity.Required.NodeSelectorTerms[i].MatchExpressions[j].Key = newKey
 			}
+			if r.Key == newKey || r.Key == oldKey {
+				for _, v := range r.Values {
+					newKeyValsSet[v] = true
+				}
+				requirementsToRemove = append(requirementsToRemove, j)
+			}
+		}
+
+		// Combine any duplicated keys
+		if len(requirementsToRemove) > 1 {
+			// This means there are duplicated new Labels, we shall combine them together
+			for j := len(requirementsToRemove) - 1; j >= 0; j-- {
+				index := requirementsToRemove[j]
+				ncRequirements := pv.Spec.NodeAffinity.Required.NodeSelectorTerms[i].MatchExpressions
+				ncRequirements[len(ncRequirements)-1], ncRequirements[index] = ncRequirements[index], ncRequirements[len(ncRequirements)-1]
+				pv.Spec.NodeAffinity.Required.NodeSelectorTerms[i].MatchExpressions = ncRequirements[:len(ncRequirements)-1]
+			}
+			// after remove all, append the new value
+			var newKeyVals []string
+			for newKeyVal := range newKeyValsSet {
+				newKeyVals = append(newKeyVals, newKeyVal)
+			}
+			sort.Strings(newKeyVals)
+			pv.Spec.NodeAffinity.Required.NodeSelectorTerms[i].MatchExpressions =
+				append(pv.Spec.NodeAffinity.Required.NodeSelectorTerms[i].MatchExpressions, v1.NodeSelectorRequirement{
+					Key:      newKey,
+					Operator: v1.NodeSelectorOpIn,
+					Values:   newKeyVals,
+				})
 		}
 	}
+
 	return nil
 }
 
-// getTopologyZones returns all topology zones with the given key found in the PV.
-func getTopologyZones(pv *v1.PersistentVolume, key string) []string {
+// getTopologyValues returns all unique topology values with the given key found in
+// the PV NodeAffinity. Sort by alphabetical order
+func getTopologyValues(pv *v1.PersistentVolume, key string) []string {
 	if pv.Spec.NodeAffinity == nil ||
 		pv.Spec.NodeAffinity.Required == nil ||
 		len(pv.Spec.NodeAffinity.Required.NodeSelectorTerms) < 1 {
 		return nil
 	}
 
-	var values []string
+	values := make(map[string]bool)
 	for i := range pv.Spec.NodeAffinity.Required.NodeSelectorTerms {
 		for _, r := range pv.Spec.NodeAffinity.Required.NodeSelectorTerms[i].MatchExpressions {
 			if r.Key == key {
-				values = append(values, r.Values...)
+				for _, v := range r.Values {
+					values[v] = true
+				}
 			}
 		}
 	}
-	return values
+	var re []string
```

Comment:
Hmm, because values is a map so that we can get rid of duplicate values.
There will be cases in different terms we have same key, value so there will be duplication. Sort is because we will probably use the slice to generate PV label and we want that in order.

Reply:
Ok, that seems reasonable. Can you add some comments explaining those?

Also it may be worth commenting that this function collapses multiple zones into a list that is ORed. It assumes the plugin doesn't support a constraint like "zone1" and "zone2"

Reply:
Done